sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -218,6 +218,44 @@ struct ImageDeleter {
   }
 };
 
+static bool is_compiler_available(const context &C,
+                                  RT::PiDeviceBinaryType Format) {
+  // No compilation needed for native format
+  if (Format == PI_DEVICE_BINARY_TYPE_NATIVE)
+    return true;
+
+  // All other formats need a compiler to be available
+  auto P = C.get_platform();
+  bool found_compiler = false;
+  for (const auto &D : C.get_devices()) {
+    if (D.get_info<info::device::is_compiler_available>()) {
+      found_compiler = true;
+      break;
+    }
+  }
+  if (!found_compiler)
+    return false;
+
+  // SPIR-V additionally needs to be able to compile IL
+  if (Format == PI_DEVICE_BINARY_TYPE_SPIRV) {
+    // OpenCL 2.1 and greater require clCreateProgramWithIL
+    if (P.get_info<info::platform::version>() >= "2.1")
```

Comment:
I think this is missing a check the platform is OpenCL.
I can't think of anything better than piUseBackend at the moment.

Reply:
Do you want this as:
if (piUseBackend(SYCL_BE_PI_OPENCL) && P.get_info<info::platform::version>() >= "2.1")) ?

Or do you want to return **true** if !piUseBackend(SYCL_BE_PI_OPENCL)?  I am assuming that a non-opencl platform would be handled in piProgramCreate?

Reply:
I think it should be former. Also I think it would be good to extend platform's get_info query to return underlying BE.

Reply:
Is extending the platform get_info something you expect me to do?

Reply:
Yes, it would be good to polish our interfaces when need arises.

Reply:
Since I know very little about the SYCL implementation, it might be better if someone else did it.  If you would like me to do it, can I get some more details about what you would like me to implement?  I am guessing that you want P.get_info<info::platform::backend> to return something.  string_class? enum pi::PiBackend?  What should the host version return?

Reply:
If you feel it holds this PR, then please go ahead without this change. It can certainly go by itself. Yes, I think `P.get_info<info::platform::backend>`  returning pi::PiBackend is a good choice. We could add a new enum value for the host platform. @bader , what do you think?

Reply:
I would prefer to put in the change without this PR.

Reply:
OK, fine with me.

Reply:
I agree with Mark.
We should complete migration of OpenCL API to Plugin Interface as a separate task.